crates/wasi/src/preview2/ip_name_lookup.rs

```diff
@@ -0,0 +1,138 @@
+use crate::preview2::bindings::sockets::ip_name_lookup::{Host, HostResolveAddressStream};
+use crate::preview2::bindings::sockets::network::{
+    Error, ErrorCode, IpAddress, IpAddressFamily, Network,
+};
+use crate::preview2::poll::{subscribe, Pollable, Subscribe};
+use crate::preview2::{AbortOnDropJoinHandle, WasiView};
+use anyhow::Result;
+use std::io;
+use std::mem;
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::pin::Pin;
+use std::vec;
+use wasmtime::component::Resource;
+
+pub enum ResolveAddressStream {
+    Waiting(AbortOnDropJoinHandle<io::Result<Vec<IpAddress>>>),
+    Done(io::Result<vec::IntoIter<IpAddress>>),
+}
+
+#[async_trait::async_trait]
+impl<T: WasiView> Host for T {
+    fn resolve_addresses(
+        &mut self,
+        network: Resource<Network>,
+        name: String,
+        family: Option<IpAddressFamily>,
+        include_unavailable: bool,
+    ) -> Result<Resource<ResolveAddressStream>, Error> {
+        let network = self.table().get_resource(&network)?;
+
+        // `Host::parse` serves us two functions:
+        // 1. validate the input is not an IP address,
+        // 2. convert unicode domains to punycode.
+        let name = match url::Host::parse(&name).map_err(|_| ErrorCode::InvalidName)? {
+            url::Host::Domain(name) => name,
+            url::Host::Ipv4(_) => return Err(ErrorCode::InvalidName.into()),
+            url::Host::Ipv6(_) => return Err(ErrorCode::InvalidName.into()),
+        };
+
+        if !network.allow_ip_name_lookup {
+            return Err(ErrorCode::PermanentResolverFailure.into());
+        }
+
+        // ignored for now, should probably have a future PR to actually take
+        // this into account. This would require invoking `getaddrinfo` directly
+        // rather than using the standard library to do it for us.
+        let _ = include_unavailable;
+
+        // For now use the standard library to perform actual resolution through
+        // the usage of the `ToSocketAddrs` trait. This blocks the current
+        // thread, so use `spawn_blocking`. Finally note that this is only
+        // resolving names, not ports, so force the port to be 0.
+        let task = tokio::task::spawn_blocking(move || -> io::Result<Vec<_>> {
```

Comment:
Should we add a `preview2::spawn_blocking` function that returns an `AbortOnDropJoinHandle`?